timemachine/cpp/src/kernels/k_nonbonded_common.cuh

```diff
@@ -12,23 +12,157 @@ static const int NONBONDED_KERNEL_THREADS_PER_BLOCK = 256;
 #define PI 3.141592653589793115997963468544185161
 #define TWO_OVER_SQRT_PI 1.128379167095512595889238330988549829708
 
-double __device__ __forceinline__ real_es_factor(double real_beta, double dij, double inv_d2ij, double &erfc_beta_dij) {
-    double beta_dij = real_beta * dij;
+// f64 switch fxn
+double __device__ __forceinline__ switch_fn(double dij) {
+    // constants
+    constexpr double cutoff = 1.2;
+    constexpr double inv_cutoff = 1 / cutoff;
+    constexpr double pi = static_cast<double>(PI);
+
+    // (dij/cutoff)^8
+    double dij_k = dij * inv_cutoff;
+    double dij_k2 = dij_k * dij_k;
+    double dij_k4 = dij_k2 * dij_k2;
+    double dij_k8 = dij_k4 * dij_k4;
+
+    // cos(0.5*pi*(dij/cutoff)^8)^3
+    double cos_arg = cos(0.5 * pi * dij_k8);
+    double cos_arg3 = cos_arg * cos_arg * cos_arg;
+    return cos_arg3;
+}
+
+double __device__ __forceinline__ d_switch_fn_dr(double dij) {
+    // constants
+    constexpr double cutoff = 1.2;
+    constexpr double pi = static_cast<double>(PI);
+
+    // cutoff^-8
+    constexpr double inv_cutoff = 1.0 / cutoff;
+    constexpr double k2 = inv_cutoff * inv_cutoff;
+    constexpr double k4 = k2 * k2;
+    constexpr double k8 = k4 * k4;
+
+    // dij^-7 and dij^-8
+    double dij2 = dij * dij;
+    double dij4 = dij2 * dij2;
+    double dij7 = dij4 * dij2 * dij;
+    double dij8 = dij4 * dij4;
+
+    // arg = 0.5 * pi * (dij / cutoff)^8
+    double dij_k8 = dij8 * k8;
+    double arg = 0.5 * pi * dij_k8;
+
+    double sin_arg;
+    double cos_arg;
+    sincos(arg, &sin_arg, &cos_arg);
+
+    double cos_arg2 = cos_arg * cos_arg;
+    constexpr double minus_12_pi_k8 = -12 * pi * k8;
+
+    return dij7 * sin_arg * cos_arg2 * minus_12_pi_k8;
+}
+
+double __device__ __forceinline__ d_erfc_beta_r_dr(double beta, double dij) {
+    // -2 beta exp(-(beta dij)^2) / sqrt(pi)
+    double beta_dij = beta * dij;
     double exp_beta_dij_2 = exp(-beta_dij * beta_dij);
-    erfc_beta_dij = erfc(beta_dij);
-    return -inv_d2ij * (static_cast<double>(TWO_OVER_SQRT_PI) * beta_dij * exp_beta_dij_2 + erfc_beta_dij);
+    return -static_cast<double>(TWO_OVER_SQRT_PI) * beta * exp_beta_dij_2;
 }
 
-float __device__ __forceinline__ real_es_factor(float real_beta, float dij, float inv_d2ij, float &erfc_beta_dij) {
```

Comment:
nit: Would it make sense to leave the existing ("reference") functional form in place to allow for tests asserting that e.g. the switched form agrees with the original when the cutoff is large?